crates/sui-framework/src/natives/types.rs

```diff
@@ -20,19 +23,50 @@ pub fn is_one_time_witness(
     debug_assert!(args.len() == 1);
 
     // unwrap safe because the interface of native function guarantees it.
-    let type_tag = context.type_to_type_tag(&ty_args.pop().unwrap())?;
+    let ty = ty_args.pop().unwrap();
+    let type_tag = context.type_to_type_tag(&ty)?;
+    let type_layout = context.type_to_type_layout(&ty)?;
 
     // TODO: what should the cost of this be?
     let cost = legacy_length_cost();
+    let struct_layout = match type_layout {
+        Some(MoveTypeLayout::Struct(l)) => l,
+        _ => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
+    };
+
+    let has_one_bool_field = match struct_layout {
+        MoveStructLayout::Runtime(vec) => {
+            if !vec.len() == 1 {
+                false
+            } else {
+                matches!(vec[0], MoveTypeLayout::Bool)
```

Comment:
You might be able to do `matches!(vec.as_slice(), [MoveTypeLayout::Bool])`

Reply:
Rust's cleverness never fails to surprise me! Thanks!

Reply:
The slice pattern matches are relatively new :) So probably won't see them around too too much